### ISSUE TITLE
Database joins (epic)

### ISSUE THREAD

**Issue by Isakdl** [serverpod/serverpod#1145]
**Is your feature request related to a problem? Please describe.**

Full support for 1:1, 1:n n:m and self relationships that can be queried automatically from the generated code. 

### Dependencies:
- [x] https://github.com/serverpod/serverpod/issues/1149

### General:
- [x] https://github.com/serverpod/serverpod/issues/1144
- [x] https://github.com/serverpod/serverpod/issues/1148
- [x] https://github.com/serverpod/serverpod/issues/1150
- [x] https://github.com/serverpod/serverpod/issues/1231
- [x] https://github.com/serverpod/serverpod/issues/1142
- [x] https://github.com/serverpod/serverpod/issues/1267
- [x] https://github.com/serverpod/serverpod/issues/1281
- [x] https://github.com/serverpod/serverpod/issues/1312
- [x] https://github.com/serverpod/serverpod/issues/1342
- [x] https://github.com/serverpod/serverpod/issues/1347
- [x] https://github.com/serverpod/serverpod/issues/1381
- [x] https://github.com/serverpod/serverpod/issues/1475
- [ ] https://github.com/serverpod/serverpod/issues/1147
- [x] https://github.com/serverpod/serverpod/issues/1232
- [ ] https://github.com/serverpod/serverpod/issues/1237
- [ ] https://github.com/serverpod/serverpod/issues/1238
- [x] https://github.com/serverpod/serverpod/issues/1257
- [x] https://github.com/serverpod/serverpod/issues/1321
- [ ] https://github.com/serverpod/serverpod/issues/1343
- [x] https://github.com/serverpod/serverpod/issues/1485
- [x] https://github.com/serverpod/serverpod/issues/1486
- [x] https://github.com/serverpod/serverpod/issues/1422
- [x] https://github.com/serverpod/serverpod/issues/1487
- [x] https://github.com/serverpod/serverpod/issues/1488
- [x] #1506 

### Support 1:1 relation ✅
- [x] https://github.com/serverpod/serverpod/issues/1293
- [x] https://github.com/serverpod/serverpod/issues/1290
- [x] https://github.com/serverpod/serverpod/issues/1151
- [x] https://github.com/serverpod/serverpod/issues/1340
### Support 1:n relation ✅
- [x] https://github.com/serverpod/serverpod/issues/1407
- [x] https://github.com/serverpod/serverpod/issues/1307
  - [x] https://github.com/serverpod/serverpod/issues/1403
  - [x] https://github.com/serverpod/serverpod/issues/1404
  - [x] https://github.com/serverpod/serverpod/issues/1405
  - [x] https://github.com/serverpod/serverpod/issues/1406
- [x] https://github.com/serverpod/serverpod/issues/1308
- [x] https://github.com/serverpod/serverpod/issues/1154
- [x] https://github.com/serverpod/serverpod/issues/1155
- [x] https://github.com/serverpod/serverpod/issues/1239
- [x] https://github.com/serverpod/serverpod/issues/1341
- [x] https://github.com/serverpod/serverpod/issues/1400
### Support m:n relation
- [x] https://github.com/serverpod/serverpod/issues/1156
- [ ] https://github.com/serverpod/serverpod/issues/1157
- [x] https://github.com/serverpod/serverpod/issues/1158
### Support self-relations  ✅
- [x] https://github.com/serverpod/serverpod/issues/1489

**Comment by lukehutch** [comment]
@SandPod I'm continuing the discussion from #1399 here, since you closed that issue, so I can no longer comment...

--

@Isakdl Thank you for the very detailed response, that was very helpful! I hope you can make the documentation this detailed when you push out a release :-)

One thing that is missing is an example of how to access the joined record in the query result. I assume, for your first example, it is like this:

```dart
final result = await Citizen.db.findSingleRow(
  session,
  where: (t) => Citizen.t.id == citizenId,
  include: Citizen.include(
    address: Address.include(),
  ),
);
String? citizenAddress = result?.address;
```

in other words, there is an `address` field added to `Citizen` because of the `relation` directive? Is this correct? Is the `result.address` field `null` when you don`t use `include` to perform the join?

One question I have is why the relation has to be specified on both sides of the relation. For a 1:1 relationship where only ever need to join in one direction, not in the reverse direction, it seems like you should only need to specify the relationship on the originating side of the join. (Although in that case, probably the tables should be combined into one table anyway -- that's a natural conclusion for 1:1 relationships!)

In general, in my opinion it is more helpful to think about joins as being injective, surjective, or bijective, than it is to think about joins being 1:1, 1:many, or many:many. (So, for example, bijective joins should be allowed, but they hint that two tables should probably be combined into one.)

Does it help reason through the join problem if you specifically think of joins as being either injective, surjective, or bijective?

For example, if a join is surjective, then there is (probably?) no sense in specifying the inverse relation, because the relation is not invertible... or if a join is injective, then the key of the target table for the join has to be unique, but it can be a subset of the set of keys for the origin table for the join.

What does it mean for a relation to be unnamed? `author: Author?, relation` -- will the relation be called `author_author`? Actually, in this case I don't see a reverse relation defined in the `author` table...

What does this mean, exactly?

```dart
where: (t) => t.author.blockedBy((b) => b.blockee.notEquals(reader.id))
```

Normally a `where` clause needs to have a Boolean type. This type is a result set of type `List<Blocked>?`. Should it therefore be the following?:

```dart
where: (t) => t.author.blockedBy((b) => b.blockee.notEquals(reader.id)).isEmpty
```

I noticed you're automatically inferring the name of a relation in some places (e.g. `post_author`) by concatenating the table names with an underscore. I assume that you only have to specify the name of a relation if there are two different types of relationship with a table, for example `blocker` and `blocked_by`. In those cases, wouldn't it be better to specify in the originating table (in this case, `author`), what the foreign key field is in the other table, rather than naming the relation on both sides, more like the actual target join syntax? e.g.:

```yaml
# blocked.yaml
class: Blocked
table: blocked
fields:
  blocker: int   # Changed
  blockee: int   # Changed
indexes:
  blocked_blocker_blockee_idx:
    fields: blocker, blockee
    unique: true

# author.yaml
class: Author
table: author
fields:
  name: String
  blockedBy: List<Blocked>?, relation(key=id, foreignKey=blocked.blockee)    # Changed
  blocked: List<Blocked>?, relation(key=id, foreignKey=blocked.blocker)    # Changed
```

Why does the `parent=author` part have to be specified at all? This all seems very foreign from an SQL perspective...

**Comment by lukehutch** [comment]
@SandPod Just to make my example more complete, and to possibly stretch your query infrastructure further, I want to be able to say "Return all posts that have an author X that has neither blocked, nor been blocked by, another user Y". (The "neither..nor" part is the addition.)

**Comment by lukehutch** [comment]
I added a StackOverflow question, because to be honest, I can't even figure out how to do this efficiently in SQL. Maybe eventually there will be some inspiration here:

https://stackoverflow.com/questions/77204668/how-to-implement-a-postgres-join-against-a-user-block-list

**Comment by lukehutch** [comment]
OK, it looks like the query I need is something like:

```sql
select p.* from posts p
    where not exists (select 1 from blocked b
        where ((b.blockee = p.author and b.blocker = $VIEWER)
            or (b.blockee = $VIEWER and b.blocker = p.author)));
```

**Comment by lukehutch** [comment]
This actually makes me wonder if queries should be split into two stages: (1) a query builder, and (2) execution of the query. That way, queries could be combined using sub-selects like this:

```dart
PostQuery postQuery = Posts.select(whereNotExists: (p) =>
    Blocked.select(where: (b) =>
        (b.blockee.equals(p.author) & b.blocker.equals(viewer)) |
        (b.blockee.equals(viewer) & b.blocker.equals(p.author))));
List<Post> = postQuery.execute(session);
```

The current mechanism, of calling `Future<List<Table>> Table.find()`, doesn't allow for this sort of flexibility.

**Comment by Isakdl** [comment]
>  I hope you can make the documentation this detailed when you push out a release :-)

That is the idea :)

> One thing that is missing is an example of how to access the joined record in the query result ... in other words, there is an address field added to Citizen because of the relation directive? Is this correct? Is the result.address field null when you dont use include` to perform the join?

Correct. Filtering and returning data are treated as separate things/concerns. You can filter on relations without including them, include data without filtering, or both at the same time.

> One question I have is why the relation has to be specified on both sides of the relation.

You don't have too! You only need to define the side where you need the information to do operations on the data. It is perfectly fine to only define one side if that is all you need. In this case, you simply omit the `name` keyword.

> Does it help reason through the join problem if you specifically think of joins as being either injective, surjective, or bijective?

We should have support for all of the above with the current syntax.


> What does it mean for a relation to be unnamed? author: Author?, relation -- will the relation be called author_author? 

`name` is just a way for us to resolve the link between to relation fields in two different protocols. It is only used during the analysis of the protocol files to help us build the representation of the data and does not have any direct impact on the SQL code. An unnamed relation is therefore simply a relation that is only specified on one side but not the other. 

> What does this mean, exactly? ```where: (t) => t.author.blockedBy((b) => b.blockee.notEquals(reader.id))```

We are reworking this part again because we ran into a few limitations, will update again later will the new solution. But basically we are looking at introducing something like this.

```dart
where: (t) => t.author.blockedBy.any((b) => b.blockee.equals(reader.id))
where: (t) => t.author.blockedBy.every((b) => b.blockee.equals(reader.id))
where: (t) => t.author.blockedBy.none((b) => b.blockee.equals(reader.id))
where: (t) => t.author.blockedBy.count((b) => b.blockee.equals(reader.id)) > 3
```

Where `any` would mean that if any relation entry matches the statement, include it. `every` meaning that every entry needs to match the statement to be included. `none` no entry is allowed to match the statement. `count` how many statements matches the statement.

> what the foreign key field is in the other table, rather than naming the relation on both sides, more like the actual target join syntax? e.g.:

This would indeed work, our thinking here is that we want to keep the relational information as close to the current protocol file as possible, and make it as explicit as possible. Meaning by looking at a specific protocol file you should know if it has a foreign key. When you manually define a relation you always do that from the side that has the foreign key. I think this approach is helpful to help create more readable and understandable code. 

In your example, you would need to have that knowledge yourself or look at all other protocol files to understand that the fields `blockee` and `blocker` are actually foreign keys.

> Why does the parent=author part have to be specified at all? This all seems very foreign from an SQL perspective...

It has to be defined if the datatype is not another protocol file, I.E. an int (id). This is used by the analyzer to resolve the relation between the files. Otherwise, we cannot know which table we should create the relation on.

Another note here, our design direction has been to view this from the dart perspective rather than from an SQL perspective. Our goal is to create an intuitive nice interface that is powerful enough so that you can ask any reasonable SQL question. But the interface being a dart one, should feel "darty" and make sense in the object world. 

There will be tradeoffs here as the OO-model and the table model is not a perfect match.

**Comment by SandPod** [comment]
Given the new syntax this would answer your original question of "_Give me all posts where I have not blocked the author_":
``` dart
where: (posts) => posts.author.blockedBy.none((b) => b.blockee.equals(reader.id))
```

And given your extended example:
>"Return all posts that have an author X that has neither blocked, nor been blocked by, another user Y". 

``` dart
where: (posts) => 
/// Removes all posts where the author has blocked the reader
posts.author.blocked.none((b) => b.blocker.equals(reader.id)) &
/// Removes all posts where the reader has blocked the author
posts.author.blockedBy.none((b) => b.blockee.equals(reader.id))
```

**Comment by lukehutch** [comment]
@Isakdl @SandPod good work, I have a feeling this is starting to really shape up :-)

> In your example, you would need to have that knowledge yourself or look at all other protocol files to understand that the fields `blockee` and `blocker` are actually foreign keys.

OK, although to make it easy for people to understand what's going on, is there a way to take an `Expression` and see what SQL is generated from the expression? Maybe a debug mode, which prints the full SQL query to the console on every query?

> And given your extended example:

Thanks for the example code, @SandPod.

Efficiency question: does a sub-select (e.g. `where not exists (select ...)`) run more slowly than a join, e.g. does the sub-select run separately for every result row in the outer select? Or is it optimized into a join internally?

**Comment by SandPod** [comment]
> OK, although to make it easy for people to understand what's going on, is there a way to take an Expression and see what SQL is generated from the expression? Maybe a debug mode, which prints the full SQL query to the console on every query?

We actually already have a, unfortunately undocumented, feature for logging the raw SQL queries that are run towards the database. This in our insights app. First connect to the server then under Log Settings, check the box Log all queries (I recommend also checking Log all sessions).

<img width="1112" alt="Screenshot 2023-10-03 at 10 23 25" src="https://github.com/serverpod/serverpod/assets/137198655/653e74ae-50f0-4d72-a4a7-cb94dc6da34c">

Then your queries should be visible in the queries 

<img width="1112" alt="Screenshot 2023-10-03 at 10 25 46" src="https://github.com/serverpod/serverpod/assets/137198655/c3dd66b7-0aa7-441a-9216-63b93a8449f9">
window.

> Efficiency question: does a sub-select (e.g. where not exists (select ...)) run more slowly than a join, e.g. does the sub-select run separately for every result row in the outer select? Or is it optimized into a join internally?

That very much depends on the problem we are solving. Generally you would think that performing a "subquery" for each row (such as using EXISTS) would always add a big performance penalty. But since Postgres's also performs query optimization the end query usually ends up quite similar (if we are not talking about too complex cases).

Check out this article for an example of that: https://www.mssqltips.com/sqlservertip/6659/sql-exists-vs-in-vs-join-performance-comparison/

By the way, It would be really interesting knowing a bit more about what you are working on @lukehutch, if you are up for sharing, please add me on LinkedIn: https://www.linkedin.com/in/alexander-sandor-543b47229/

**Comment by SandPod** [comment]
@lukehutch, Once this, https://github.com/serverpod/serverpod/pull/1476, is in you should be able to build the expression we talked about using latest main.

``` dart
where: (posts) => 
/// Removes all posts where the author has blocked the reader
posts.author.blocked.none((b) => b.blocker.equals(reader.id)) &
/// Removes all posts where the reader has blocked the author
posts.author.blockedBy.none((b) => b.blockee.equals(reader.id))
```

**Comment by lukehutch** [comment]
@SandPod awesome, thanks!

**Comment by lukehutch** [comment]
@SandPod what other work is coming up, for queries? I need to make a decision as to when to rebase my tree on master. It's a huge amount of work, because I'm maintaining about 8 PRs against master, and they interact. So I want to make sure I rebase when things are stable, and when significant functionality has been added.

**Comment by SandPod** [comment]
> @SandPod what other work is coming up, for queries? I need to make a decision as to when to rebase my tree on master. It's a huge amount of work, because I'm maintaining about 8 PRs against master, and they interact. So I want to make sure I rebase when things are stable, and when significant functionality has been added.

We're in the final stages of preparing the 1.2 release. Here's the latest status on the remaining tasks related to queries:

- "Any filter" is already in a pull request (PR), which you can find here: [Link to PR](https://github.com/serverpod/serverpod/pull/1478).
- "Every filter" is currently a work in progress.
- For the "Many relation filters in delete queries," we might postpone this to 1.2.1, depending on its complexity.
- We're also working on some general polishing and cleanup.

Additionally, the major task preventing us from finalizing 1.2 is the UX improvements needed for database migrations. Given this, it might be advisable to wait until at least all the filters for many relations are implemented before performing the rebase. If possible it might be even better to wait until after the 1.2 release is completed.

**Comment by lukehutch** [comment]
@SandPod I checked #1478, and I don't see `parent` used there. Is that necessary?

**Comment by SandPod** [comment]
> @SandPod I checked #1478, and I don't see `parent` used there. Is that necessary?

No, when we use an Object Relation the parent can be inferred from the type of the field.

### Examples
Given the following entities:
``` company.yaml
class: Company
table: company
fields:
  name: String
  employees: List<Citizen>?, relation
```

``` citizen.yaml
class: Citizen
table: citizen
fields:
  name: String
```

We infer the parent for the `employees` field is `Citizen` because that is the type we have defined for the list.

**Comment by lukehutch** [comment]
@SandPod Oh, what is a parent relation then? I would expect the parent of `Citizen` objects to be `Company`. But then I would have had a followup question about whether a class can have multiple parents.

Maybe you could actually diagram this for the docs? Or at least show exactly what the protocol files compile into, and exactly what SQL table configs are built, and what SQL queries are built for different query types (as I suggested previously), for several different types of examples... the protocol files are not enough to understand what's going on. Thanks!

I rebased on git main, so I'll be rewriting all my queries using the new syntax as soon as I understand what's going on! Then I'll hopefully have some useful feedback for you. Some of my queries are going to be very complex.

**Comment by SandPod** [comment]
@lukehutch Great to get feedback on this!

I'm not using the correct phrasing here speaking from an SQL perspective.

What we do in the example is that we see that the `Company` class wants a many relation to the `Citizen` class. We then infeer that `Citizen` needs to have `Company` as parent and create a field in the `Citizen` table to hold a foreign key to `Company`.

But since we have not specified that we want this field accessible in `Citizen` in the protocol we don't generate any fields in the dart code for accessing the relationship from that side.

If we want the relation to be accessible (as an object relation enabling relational queries) from both side we need to specify a named relation like so:
``` company.yaml
class: Company
table: company
fields:
  name: String
  employees: List<Citizen>?, relation(name=company_employees)
```

``` citizen.yaml
class: Citizen
table: citizen
fields:
  name: String
  company: Company?, relation(name=company_employees)
```

The idea is that when you specify your protocol files you should come as close as possible to only thinking in a model/relationship perspective. What fields do I want available on my dart class and what kind of relationships do I want these entities to have to other entities?

**Comment by SandPod** [comment]
If we don't want an object relation to be accessible on the `Citizen` class but we would like to have access to the `Id` of the company related to it, we could also specify the relationship like so:
``` company.yaml
class: Company
table: company
fields:
  name: String
  employees: List<Citizen>?, relation(name=company_employees)
```

``` citizen.yaml
class: Citizen
table: citizen
fields:
  name: String
  companyId: int?, relation(name=company_employees)
```

This is basically the same relation we end up with in the first example where we only specified the relation on the `Company` side.

**Comment by SandPod** [comment]
@lukehutch If you want, you could actually spin upp our WIP documentation for all of this on this branch here: https://github.com/serverpod/serverpod_docs/tree/database

Pretty simple if you have npm installed.

Then you just navigate to the `next` version in the documentation interface.

Would love any feedback on it, but just remember that it is still WIP.

**Comment by lukehutch** [comment]
> @lukehutch, Once this, #1476, is in you should be able to build the expression we talked about using latest main.
> 
> ```dart
> where: (posts) => 
> /// Removes all posts where the author has blocked the reader
> posts.author.blocked.none((b) => b.blocker.equals(reader.id)) &
> /// Removes all posts where the reader has blocked the author
> posts.author.blockedBy.none((b) => b.blockee.equals(reader.id))
> ```

OK, I ran into an issue with this, post #1476:

I have tables

```yaml
class: Endorsement
table: endorsement
fields:
  ### The endorsing user
  endorserUserId: int
  ### The endorsed user
  endorseeUserId: int
  howKnow: String?
  endorsementText: String
  approved: bool
  date: DateTime
  likeActions: List<LikeAction>?, relation
indexes:
  endorsement_endorser_user_id_idx:
    fields: endorserUserId
  endorsement_endorsee_user_id_idx:
    fields: endorseeUserId

class: LikeAction
table: like_action
fields:
  ### The initiating user
  userId: int
  ### The target user
  otherUserId: int
  likeActionType: LikeActionType
  modified: DateTime
indexes:
  like_action_user_id_idx:
    fields: userId
  like_action_other_user_id_idx:
    fields: otherUserId
  like_action_modified_idx:
    fields: modified

enum: LikeActionType
values:
  - like
  - dislike
  - block
```

Now I need to write a query that shows all endorsements for another user:

```dart
  Future<List<Endorsement>?> getEndorsements(
    Session session,
    int endorseeUserId,
  ) {
    final viewerUserId = (await session.auth.authenticatedUserId)!;
    return Endorsement.db.find(session,
        where: (t) =>
            t.endorseeUserId.equals(endorseeUserId) &
            // Return all endorsements for own profile, even if not approved
            (Constant.bool(viewerUserId == endorseeUserId) |
                // In general only return approved endorsements if viewing
                // endorsements on another user's profile
                t.approved.equals(true) |
                // However, return even unapproved endorsement if the viewer
                // is the endorser
                t.endorserUserId.equals(viewerUserId)));
  }
```

Now I want to extend this query to exclude any endorsements, given the `LikeAction` table, where `LikeActionType` is `blocked`. This needs to be applied 3-ways between viewer, endorser, and endorsee.

Here is how I implemented this query (with the goal of appending `& _notBlocked(...)` to the end of the `getEndorsements` query):

```dart
  Expression _notBlocked({
    required EndorsementTable endorsementTable,
    required int endorserUserId,
    required int endorseeUserId,
    required int viewerUserId,
  }) {
    // Check 2-way block between endorser and endorsee
    Expression endorsementCheck(LikeActionTable likeAction) =>
        (likeAction.userId.equals(endorserUserId) &
            likeAction.otherUserId.equals(endorseeUserId)) |
        (likeAction.userId.equals(endorseeUserId) &
            likeAction.otherUserId.equals(endorserUserId));
    return endorsementTable.likeActions.none((likeAction) =>
        (viewerUserId == endorseeUserId || viewerUserId == endorseeUserId
            // 2-way check
            ? endorsementCheck(likeAction)
            // If viewer is a 3rd party, then need to check all 6 block
            // directions (3-way check)
            : endorsementCheck(likeAction) &
                ((likeAction.userId.equals(viewerUserId) &
                        likeAction.otherUserId.equals(endorseeUserId)) |
                    (likeAction.userId.equals(endorseeUserId) &
                        likeAction.otherUserId.equals(viewerUserId)) |
                    (likeAction.userId.equals(viewerUserId) &
                        likeAction.otherUserId.equals(endorserUserId)) |
                    (likeAction.userId.equals(endorserUserId) &
                        likeAction.otherUserId.equals(viewerUserId)))) &
        likeAction.likeActionType.equals(LikeActionType.block));
  }
```

The problem is that this block-checking expression requires a runtime-provided `endorserUserId`, whereas the query in `getEndorsements` returns all endorsements for the requested `endorseeUserId`, regardless of the `endorserUserId`.

I need to be able to modify `_notBlocked` so that it can join `Endorsement.endorserUserId` to `LikeActions.userId` or `LikeActions.otherUserId`, anywhere where `_notBlocked` uses the integer value `endorserUserId`.

I don't see a way to perform a join of this kind, using the current join syntax. Please advise.

**Comment by lukehutch** [comment]
I have another issue. When I define

```yaml
class: Endorsement
table: endorsement
fields:
  # ...
  likeActions: List<LikeAction>?, relation
```

then Serverpod creates a column in the `like_action` table called `_endorsementLikeactionsEndorsementId`. However, I don't want `Endorsement` to actually be a parent relation of `LikeAction`! I just want to be able to join between `Endorsement` and `LikeAction`, i.e. fetch any relevant `LikeAction` rows for each result in a query that returns `Endorsement` rows.

Let me explain this with a concrete example:

* User A is viewing the profile of User C.
* User C has been endorsed by User B.
  * So on the profile view of User C, viewed by User A, the following will be displayed: "User B endorsed User C", along with User B's avatar icon.
* User A previously liked User B, meaning there is a `LikeAction` row with User A as `userId`, User B as `otherId`, and `LikeActionType.like` as `likeActionType`.
  * When the endorsement of User C by User B is fetched, `Endorsement.likeActions` should contain the above `LikeAction` row.

The problem is that the `LikeAction` rows are not conceptually nested within `Endorsement` -- they are a completely separate set of entities.

Actually we have a 3-level query here:

* `Profile`, for a specific `Profile.userId`
  * `List<Endorsement>` where `Endorsement.endorseeUserId == Profile.userId`
    * `List<LikeAction>` where `LikeAction.userId.equals($viewerUserId) & LikeAction.otherUserId.equals(Endorsement.endorserUserId)`

I don't want "back-link" columns from `LikeAction` to `Endorsement`, although I guess it's OK for `Profile` to be the parent of `Endorsement`, since endorsements do conceptually belong to a profile.

**Comment by lukehutch** [comment]
Actually maybe I'm misunderstanding the purpose of `relation` fields. I see that the relation field is actually present in result rows, not just in the query syntax. I think I now see why you are calling them `1:n` relationships: the records in the joined table can have only one parent, so every child table row belongs to exactly one parent table row?

How does Postgres return `n` child records for one parent record? I can see how they can be used as part of a sub-select returning a boolean value, but I thought that SQL result sets can only return rectangular result sets, so does Postgres just duplicate all the parent fields for each child record, then you de-dup this in Serverpod?

**Comment by Isakdl** [comment]
> the records in the joined table can have only one parent, so every child table row belongs to exactly one parent table row?

Yes! The only real difference between a `1:1` and a `1:n` relation is that a `1:1` has a unique index on the foreign key. But yes to create a relation you need to have a foreign key stored in one of the two tables you are trying to link. 

> How does Postgres return n child records for one parent record? I can see how they can be used as part of a sub-select returning a boolean value, but I thought that SQL result sets can only return rectangular result sets, so does Postgres just duplicate all the parent fields for each child record, then you de-dup this in Serverpod?

Our solution here for 1:n relations when `including` the data in the response (which is different from filtering), is to make 1 additional query per `ListInclude` that is used. This allows us to support the full feature set of filtering, ordering, and paginating the sub-list.

```dart
var endorsements = Endorsement.db.find(
  session,
  include: Endorsement.include(
    likeActions: LikeAction.includeList(
      where: (t) => ...,
      orderBy: ...,
      limit: ...,
      offset: ...,
    ),
  ),
);
```

**Comment by lukehutch** [comment]
@Isakdl I couldn't figure out how this could work without running a separate query, so I went looking for the code that run an additional query per `ListInclude`, but I didn't end up finding it -- hence my puzzlement!

If that's the case, I should just run those queries myself -- that way I get more control anyway.

Sorry for the long and complex messages I left earlier, but please also respond to my question about adding `.equalsColumn`. I think this really needs to be supported (even if I'm unlikely to use these parent/child-relationships).

**Comment by Isakdl** [comment]
Closing as completed, we track the remaining issues separately and will build them in isolation outside of this epic.
